transport_test.go

```
@@ -42,3 +57,143 @@ func TestInitTransportWithCustomCA(t *testing.T) {
+
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("failed to listen on secure address: %v", err)
+	}
+	defer l.Close()
+	tlsListener := tls.NewListener(l, &tls.Config{
+		Certificates: []tls.Certificate{tlsCert},
+		ClientCAs:    clientCA,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+	})
+	defer tlsListener.Close()
+
+	go func() {
+		if err := tlsServer.Serve(tlsListener); err != nil {
+			t.Logf("failed to run the test server: %v", err)
+		}
+	}()
+	defer tlsServer.Close()
+
+	tmpDir := t.TempDir()
+	serverCertPath := filepath.Join(tmpDir, "server.crt")
+	clientCertPath := filepath.Join(tmpDir, "client.crt")
+	clientKeyPath := filepath.Join(tmpDir, "client.key")
+
+	if err := certutil.WriteCert(serverCertPath, cert); err != nil {
+		t.Fatalf("failed to write server cert: %v", err)
+	}
+	if err := certutil.WriteCert(clientCertPath, clientCert); err != nil {
+		t.Fatalf("failed to write client cert: %v", err)
+	}
+	if err := keyutil.WriteKey(clientKeyPath, clientKey); err != nil {
+		t.Fatalf("failed to write client key: %v", err)
+	}
+
+	roundTripper, err := initTransport(serverCertPath, clientCertPath, clientKeyPath)
+	if err != nil {
+		t.Errorf("want err to be nil, but got %v", err)
+		return
+	}
+
+	httpReq, err := http.NewRequest(http.MethodPost, fmt.Sprintf("https://127.0.0.1:%d", l.Addr().(*net.TCPAddr).Port), nil)
+	if err != nil {
+		t.Fatalf("failed to create an HTTP request: %v", err)
+	}
+
+	resp, err := roundTripper.RoundTrip(httpReq)
+	if err != nil {
+		t.Fatalf("request failed: %v", err)
+	}
+	defer resp.Body.Close()
+
+	select {
+	case isCorrectRequest := <-requestReceived:
+		if !isCorrectRequest {
+			t.Errorf("the server did not receive the expected request")
+		}
+	case <-time.NewTimer(1 * time.Second).C:
+		t.Fatalf("the server did not receive any requests in the required time")
+	}
+
+}
+
+func generateClientCert(t *testing.T) ([]byte, []byte, *x509.CertPool) {
```

It would be better to return an error and handle the error case with one `t.Fatal` in the test itself.